Add support for using nested travel method

meinac opened the pull request:
One of my friend saw that while working on an issue about rubygems, it would be great if we can use nested `travel` methods and asked me to implement this.
By this way we can use nested `travel` methods in our tests. For example;

``` ruby
travel 1.day do
  Foo.do_something #This method depends on Time.now and it has mock value lets say A

  travel 1.day do
    Bar.do_another_thing #This method depends on Time.now and it has mock value lets say B
  end

  Zoo.do_yet_another_thing #This method depends on Time.now and it has mock value as A
end
```

In current implementation we can't use nested `travel` methods because when the execution of deepest `travel` method finishes `Time.now` returns back to original `Time.now`.
Lets see this in an example like above;

``` ruby
travel 1.day do
  Foo.do_something #This method depends on Time.now and it has mock value lets say A

  travel 1.day do
    Bar.do_another_thing #This method depends on Time.now and it has mock value lets say B
  end

  Zoo.do_yet_another_thing #This method depends on Time.now and it returns the original Time.now
end
```

As you can see we've expected that `Time.now` will return `A` but it returns original `Time.now`.

meinac committed:
Add support for using nested travel method

sgrif commented on activesupport/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+*   Added support for using nested travel methods
```
This should be explained further.

sgrif commented on activesupport/lib/active_support/testing/time_helpers.rb:
```diff
@@ -30,13 +38,27 @@ def unstub_all!
       end
 
       private
-
```
Please leave this line in.

sgrif commented on activesupport/lib/active_support/testing/time_helpers.rb:
```diff
@@ -30,13 +38,27 @@ def unstub_all!
+          key = [object.object_id, method_name]
+
+          if stub = @stubs[key]
+            old_stubs = @pre_stubs.delete(key){ Array.new }
```
Add a space before the `{`

sgrif commented on activesupport/lib/active_support/testing/time_helpers.rb:
```diff
@@ -30,13 +38,27 @@ def unstub_all!
+            old_stubs << {
+              object: object,
+              method_name: method_name,
+              return_value: object.send(method_name)
```
Add a trailing comma

sgrif replied:
This should probably use `public_send`

sgrif commented on activesupport/lib/active_support/testing/time_helpers.rb:
```diff
@@ -30,13 +38,27 @@ def unstub_all!
+              method_name: method_name,
+              return_value: object.send(method_name)
+            }
+            @pre_stubs.merge!(key => old_stubs)
```
Why are we just deleting and then merging? Why not just `@pre_stubs[key] ||= []; @pre_stubs[key] << ...`

sgrif commented on activesupport/lib/active_support/testing/time_helpers.rb:
```diff
@@ -22,6 +21,15 @@ def stub_object(object, method_name, return_value)
+        key = [object.object_id, method_name]
+        unstub_object(@stubs.delete(key))
+
+        if old_stub = @pre_stubs.fetch(key, Array.new).pop
```
Use `[]` instead of `Array.new`

sgrif replied:
I find this very hard to understand. Maybe extract this line into a method to clarify its intent?

meinac commented:
@sgrif what about this?

maclover7 commented:
Please rebase.

meinac commented:
Hi @maclover7, thank you to take a look at this. Nowadays I couldn't take a look into Rails issues because of my huge job load and I'm really unhappy with this situation. I've rebased this and it's ready to merge into master, thanks.

maclover7 commented:
Are we ready to roll here, @sgrif?

nateberkopec commented:
FWIW I still don't quite understand the changelog entry. It could be clarified.

meinac commented:
I've changed the changelog entry and added examples about old and new behaviours to make it clear. I hope it demonstrates why this change is required.

meinac commented:
I've checked and saw that tests are in `client_test.rb` for actioncable are non-deterministic. I believe build failure is not introduced with this change because I can see same failures with master branch on my virtual machine too. I think one of the dependencies causing problem, maybe concurrent-ruby or not but I've not enough experience in this gems to dive into deep.

rafaelfranca commented:
Closing in favor of https://github.com/rails/rails/pull/24890
